File: ios/BiblosSurfer/Services/Services/ReaderSettingsStore.swift
//
//  ReaderSettingsStore.swift
//  BiblosSurfer
//
//  Created by Mikołaj Linczewski on 21/08/2026.
//

import AVFoundation
import Foundation
import ReadiumNavigator
import ReadiumShared

enum TTSChunkUnit: String, CaseIterable, Identifiable {
    case word
    case sentence
    case paragraph

    var id: String { rawValue }

    var textUnit: TextUnit {
        switch self {
        case .word: .word
        case .sentence: .sentence
        case .paragraph: .paragraph
        }
    }

    var title: String {
        switch self {
        case .word: "Word"
        case .sentence: "Sentence"
        case .paragraph: "Paragraph"
        }
    }
}

final class ReaderSettingsStore {
    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    var fontSize: Double {
        get { defaults.object(forKey: "reader.fontSize") as? Double ?? 1.0 }
        set { defaults.set(newValue, forKey: "reader.fontSize") }
    }

    var fontFamily: String {
        get { defaults.string(forKey: "reader.fontFamily") ?? "Original" }
        set { defaults.set(newValue, forKey: "reader.fontFamily") }
    }

    var theme: Theme {
        get { Theme(rawValue: defaults.string(forKey: "reader.theme") ?? "") ?? .light }
        set { defaults.set(newValue.rawValue, forKey: "reader.theme") }
    }

    var scroll: Bool {
        get { defaults.bool(forKey: "reader.scroll") }
        set { defaults.set(newValue, forKey: "reader.scroll") }
    }

    var voiceIdentifier: String? {
        get { defaults.string(forKey: "reader.voiceIdentifier") }
        set { defaults.set(newValue, forKey: "reader.voiceIdentifier") }
    }

    var speechRate: Float {
        get {
            let stored = defaults.object(forKey: "reader.speechRate") as? Float
            return stored ?? AVSpeechUtteranceDefaultSpeechRate
        }
        set { defaults.set(newValue, forKey: "reader.speechRate") }
    }

    var pitchMultiplier: Float {
        get {
            let stored = defaults.object(forKey: "reader.pitchMultiplier") as? Float
            return stored ?? 1.0
        }
        set { defaults.set(newValue, forKey: "reader.pitchMultiplier") }
    }

    var speechVolume: Float {
        get {
            let stored = defaults.object(forKey: "reader.speechVolume") as? Float
            return stored ?? 1.0
        }
        set { defaults.set(newValue, forKey: "reader.speechVolume") }
    }

    var preUtteranceDelay: TimeInterval {
        get { defaults.object(forKey: "reader.preUtteranceDelay") as? TimeInterval ?? 0 }
        set { defaults.set(newValue, forKey: "reader.preUtteranceDelay") }
    }

    var postUtteranceDelay: TimeInterval {
        get { defaults.object(forKey: "reader.postUtteranceDelay") as? TimeInterval ?? 0 }
        set { defaults.set(newValue, forKey: "reader.postUtteranceDelay") }
    }

    var defaultLanguage: String? {
        get { defaults.string(forKey: "reader.defaultLanguage") }
        set { defaults.set(newValue, forKey: "reader.defaultLanguage") }
    }

    var chunkUnit: TTSChunkUnit {
        get {
            TTSChunkUnit(rawValue: defaults.string(forKey: "reader.chunkUnit") ?? "") ?? .sentence
        }
        set { defaults.set(newValue.rawValue, forKey: "reader.chunkUnit") }
    }

    var useSystemSpeechSettings: Bool {
        get { defaults.bool(forKey: "reader.useSystemSpeechSettings") }
        set { defaults.set(newValue, forKey: "reader.useSystemSpeechSettings") }
    }

    func epubPreferences() -> EPUBPreferences {
        var preferences = EPUBPreferences()
        preferences.fontSize = fontSize
        preferences.scroll = scroll
        preferences.theme = theme
        if fontFamily != "Original" {
            preferences.fontFamily = FontFamily(rawValue: fontFamily)
            preferences.publisherStyles = false
        }
        return preferences
    }
}

File: android/app/src/main/java/miko/biblossurfer/data/ReaderSettingsStore.kt
package miko.biblossurfer.data

import android.content.SharedPreferences
import org.readium.r2.navigator.epub.EpubPreferences
import org.readium.r2.navigator.preferences.FontFamily
import org.readium.r2.navigator.preferences.Theme
import org.readium.r2.shared.util.tokenizer.TextUnit

enum class TTSChunkUnit {
    WORD,
    SENTENCE,
    PARAGRAPH;

    val textUnit: TextUnit
        get() = when (this) {
            WORD -> TextUnit.Word
            SENTENCE -> TextUnit.Sentence
            PARAGRAPH -> TextUnit.Paragraph
        }

    val title: String
        get() = when (this) {
            WORD -> "Word"
            SENTENCE -> "Sentence"
            PARAGRAPH -> "Paragraph"
        }

    val storageValue: String
        get() = when (this) {
            WORD -> "word"
            SENTENCE -> "sentence"
            PARAGRAPH -> "paragraph"
        }

    companion object {
        val allCases: List<TTSChunkUnit> = entries

        fun fromStorage(value: String?): TTSChunkUnit =
            entries.firstOrNull { it.storageValue == value } ?: SENTENCE
    }
}

class ReaderSettingsStore(
    private val defaults: SharedPreferences,
) {
    var fontSize: Double
        get() = defaults.doubleOrDefault(KEY_FONT_SIZE, 1.0)
        set(value) { defaults.edit().putDouble(KEY_FONT_SIZE, value).apply() }

    var fontFamily: String
        get() = defaults.getString(KEY_FONT_FAMILY, "Original") ?: "Original"
        set(value) { defaults.edit().putString(KEY_FONT_FAMILY, value).apply() }

    var theme: Theme
        get() = Theme.entries.firstOrNull {
            it.name.equals(defaults.getString(KEY_THEME, "") ?: "", ignoreCase = true)
        } ?: Theme.LIGHT
        set(value) { defaults.edit().putString(KEY_THEME, value.name.lowercase()).apply() }

    var scroll: Boolean
        get() = defaults.getBoolean(KEY_SCROLL, false)
        set(value) { defaults.edit().putBoolean(KEY_SCROLL, value).apply() }

    var voiceIdentifier: String?
        get() = defaults.getString(KEY_VOICE, null)
        set(value) { defaults.edit().putString(KEY_VOICE, value).apply() }

    var speechRate: Float
        get() = if (defaults.contains(KEY_SPEECH_RATE)) defaults.getFloat(KEY_SPEECH_RATE, 1.0f) else 1.0f
        set(value) { defaults.edit().putFloat(KEY_SPEECH_RATE, value).apply() }

    var pitchMultiplier: Float
        get() = if (defaults.contains(KEY_PITCH)) defaults.getFloat(KEY_PITCH, 1.0f) else 1.0f
        set(value) { defaults.edit().putFloat(KEY_PITCH, value).apply() }

    var speechVolume: Float
        get() = if (defaults.contains(KEY_VOLUME)) defaults.getFloat(KEY_VOLUME, 1.0f) else 1.0f
        set(value) { defaults.edit().putFloat(KEY_VOLUME, value).apply() }

    var preUtteranceDelay: Double
        get() = defaults.doubleOrDefault(KEY_PRE_DELAY, 0.0)
        set(value) { defaults.edit().putDouble(KEY_PRE_DELAY, value).apply() }

    var postUtteranceDelay: Double
        get() = defaults.doubleOrDefault(KEY_POST_DELAY, 0.0)
        set(value) { defaults.edit().putDouble(KEY_POST_DELAY, value).apply() }

    var defaultLanguage: String?
        get() = defaults.getString(KEY_LANGUAGE, null)
        set(value) { defaults.edit().putString(KEY_LANGUAGE, value).apply() }

    var chunkUnit: TTSChunkUnit
        get() = TTSChunkUnit.fromStorage(defaults.getString(KEY_CHUNK, null))
        set(value) { defaults.edit().putString(KEY_CHUNK, value.storageValue).apply() }

    var useSystemSpeechSettings: Boolean
        get() = defaults.getBoolean(KEY_SYSTEM_SPEECH, false)
        set(value) { defaults.edit().putBoolean(KEY_SYSTEM_SPEECH, value).apply() }

    fun epubPreferences(): EpubPreferences {
        var preferences = EpubPreferences(
            fontSize = fontSize,
            scroll = scroll,
            theme = theme,
        )
        if (fontFamily != "Original") {
            preferences = preferences.copy(
                fontFamily = FontFamily(fontFamily),
                publisherStyles = false,
            )
        }
        return preferences
    }

    companion object {
        const val KEY_FONT_SIZE = "reader.fontSize"
        const val KEY_FONT_FAMILY = "reader.fontFamily"
        const val KEY_THEME = "reader.theme"
        const val KEY_SCROLL = "reader.scroll"
        const val KEY_VOICE = "reader.voiceIdentifier"
        const val KEY_SPEECH_RATE = "reader.speechRate"
        const val KEY_PITCH = "reader.pitchMultiplier"
        const val KEY_VOLUME = "reader.speechVolume"
        const val KEY_PRE_DELAY = "reader.preUtteranceDelay"
        const val KEY_POST_DELAY = "reader.postUtteranceDelay"
        const val KEY_LANGUAGE = "reader.defaultLanguage"
        const val KEY_CHUNK = "reader.chunkUnit"
        const val KEY_SYSTEM_SPEECH = "reader.useSystemSpeechSettings"
    }
}

fun SharedPreferences.doubleOrDefault(key: String, default: Double): Double =
    if (contains(key)) java.lang.Double.longBitsToDouble(getLong(key, 0L)) else default

fun SharedPreferences.Editor.putDouble(key: String, value: Double): SharedPreferences.Editor =
    putLong(key, java.lang.Double.doubleToRawLongBits(value))
